frontend/src/app/report/histogram/histogram-chart/histogram-chart.component.ts

```diff
@@ -0,0 +1,139 @@
+import { Component, Input, OnChanges, OnInit } from '@angular/core';
+import { DecimalPipe } from "@angular/common";
+import { formatPercentage } from "@data/calculation/calculation.util";
+
+@Component({
+  selector: 'app-histogram-chart',
+  templateUrl: './histogram-chart.component.html',
+  styleUrls: ['./histogram-chart.component.scss']
+})
+
+export class HistogramChartComponent implements OnInit, OnChanges {
+  @Input() bins: number[] = []
+  @Input() reportMax!: number;
+  @Input() zeroes!: number;
+  @Input() inclusive = false;
+  @Input() locale = 'en';
+
+  readonly leftMargin = 184;
+  segments: Segment[] = [];
+  binInfo: BinInfo[] = [];
+
+  private readonly chartWidth  = 1200
+  private readonly chartHeight = 864;
+  private readonly cHeight     = 800;
+  private readonly topMargin   = 20;
+  private readonly binWidth    = 10;
+
+  private max         = 0;
+  private bmax        = 0;
+
+  constructor(private decimalPipe: DecimalPipe) { }
+
+  ngOnInit(): void {
+
+    this.bmax = Math.max(...this.bins);
+
+    const count = this.bins.reduce((a, b) => a + b)
+                + (this.inclusive ? 0 : this.zeroes),
+          binSz = this.reportMax / 100,
+          ystep =       // suitable vertical measure interval
+            this.bmax < 10 ? 1 :
+            this.bmax < 30 ? 5 :
+            this.bmax < 100 ? 10 :
+            this.bmax < 1000 ? 50 :
+            this.bmax < 3000 ? 100 :
+            this.bmax < 10000 ? 500 :
+            this.bmax < 20000 ? 1000 :
+            this.bmax < 100000 ? 5000 :
+            this.bmax < 300000 ? 10000 :
+            this.bmax < 1000000 ? 100000 :
+            500000;
+
+    let bx = 0,
+        acc = this.bins[0] + (this.inclusive ? 0 : this.zeroes),
+        yc = 0;
+
+    this.max = this.bmax < 10 ? 10 :
+      (Math.trunc(this.bmax / ystep) + 1) * ystep;
+
+    this.binInfo[0] = this.inclusive ?
+                        this.makeInfo(0, acc, count) :
+                        new BinInfo('ε', acc, count, this.zeroes, this.locale);
+
+    while (bx++ < this.bins.length -1) {
+      acc += this.bins[bx];
+      this.binInfo[bx] = this.makeInfo(bx * binSz,  acc, count);
+    }
+
+    this.binInfo[bx] = this.makeInfo(this.reportMax, acc, count);
+
+    while (yc < this.max) {
+      yc += ystep;
+      this.segments.push(
+        new Segment(this.leftMargin - this.binWidth * 3,
+          this.topMargin + this.cHeight * (1 - yc / this.max),
+          this.binWidth * 104, yc));
+    }
```

Comment:
This code does not qualify as FP ;)

Reply:
I'll take care of that :)

Reply:
You don't have to, I'm just pointing out the obvious.. FP would make it easier to understand tho..